File: ifd/ifd-core/src/main/java/org/openecard/ifd/event/IfdEventManager.java
/****************************************************************************
 * Copyright (C) 2012-2017 ecsec GmbH.
 * All rights reserved.
 * Contact: ecsec GmbH (info@ecsec.de)
 *
 * This file is part of the Open eCard App.
 *
 * GNU General Public License Usage
 * This file may be used under the terms of the GNU General Public
 * License version 3.0 as published by the Free Software Foundation
 * and appearing in the file LICENSE.GPL included in the packaging of
 * this file. Please review the following information to ensure the
 * GNU General Public License version 3.0 requirements will be met:
 * http://www.gnu.org/copyleft/gpl.html.
 *
 * Other Usage
 * Alternatively, this file may be used in accordance with the terms
 * and conditions contained in a signed written agreement between
 * you and ecsec GmbH.
 *
 ***************************************************************************/

package org.openecard.ifd.event;

import iso.std.iso_iec._24727.tech.schema.ConnectionHandleType;
import iso.std.iso_iec._24727.tech.schema.GetIFDCapabilities;
import iso.std.iso_iec._24727.tech.schema.GetIFDCapabilitiesResponse;
import iso.std.iso_iec._24727.tech.schema.IFDCapabilitiesType;
import iso.std.iso_iec._24727.tech.schema.IFDStatusType;
import iso.std.iso_iec._24727.tech.schema.Wait;
import iso.std.iso_iec._24727.tech.schema.WaitResponse;
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.atomic.AtomicInteger;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import org.openecard.common.ECardConstants;
import org.openecard.common.WSHelper;
import org.openecard.common.WSHelper.WSException;
import org.openecard.common.event.EventType;
import org.openecard.common.interfaces.Environment;
import org.openecard.common.event.IfdEventObject;
import org.openecard.common.util.HandlerBuilder;
import org.openecard.common.util.ValueGenerators;
import org.openecard.ifd.scio.wrapper.ChannelManager;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


/**
 * Main class of the event system.
 * Use this to create and operate an event manager.
 *
 * @author Tobias Wich
 * @author René Lottes
 */
public class IfdEventManager {

    private static final Logger LOG = LoggerFactory.getLogger(IfdEventManager.class);

    private static final AtomicInteger THREAD_NUM = new AtomicInteger(1);

    protected final Environment env;
    protected final ChannelManager cm;
    protected final byte[] ctx;
    protected final String sessionId;
    private final HandlerBuilder builder;

    protected ExecutorService threadPool;

    private IfdEventRunner eventRunner;
    private Future<?> watcher;


    public IfdEventManager(Environment env, ChannelManager cm, byte[] ctx) {
	this.env = env;
	this.cm = cm;
	this.ctx = ctx;
	this.sessionId = ValueGenerators.genBase64Session();
	this.builder = HandlerBuilder.create()
		.setContextHandle(ctx)
		.setSessionId(sessionId);
    }

    public synchronized void initialize() {
	threadPool = Executors.newCachedThreadPool(new ThreadFactory() {
	    private final AtomicInteger num = new AtomicInteger(0);
	    private final ThreadGroup group = new ThreadGroup("IFD Event Manager");
	    @Override
	    public Thread newThread(Runnable r) {
		String name = String.format("IFD Watcher %d", num.getAndIncrement());
		Thread t = new Thread(group, r, name);
		t.setDaemon(false);
		return t;
	    }
	});
	// start watcher thread
	try {
	    eventRunner = new IfdEventRunner(env, this, builder, cm, ctx);
	    watcher = threadPool.submit(eventRunner);
	} catch (WSException ex) {
	    throw new RuntimeException("Failed to request initial status from IFD.");
	}
    }

    public synchronized void terminate() {
	eventRunner.setStoppedFlag();
	watcher.cancel(true);
	threadPool.shutdownNow();
    }

    @Nonnull
    protected List<IFDStatusType> wait(@Nonnull List<IFDStatusType> lastKnown) throws WSException {
	Wait wait = new Wait();
	wait.setContextHandle(ctx);
	wait.getIFDStatus().addAll(lastKnown);
	WaitResponse resp = env.getIFD().wait(wait);

	try {
	    WSHelper.checkResult(resp);
	    List<IFDStatusType> result = resp.getIFDEvent();
	    return result;
	} catch (WSException ex) {
	    if (ECardConstants.Minor.IFD.INVALID_SLOT_HANDLE.equals(ex.getResultMinor())) {
		// this can only happen when the PCSC stack is reloaded, notify all cards have disappeared
		ArrayList<IFDStatusType> result = new ArrayList<>(lastKnown.size());
		if (! lastKnown.isEmpty()) {
		    LOG.info("PCSC stack seemed to disappear. Signalling that no cards are available anymore.");
		    for (IFDStatusType next : lastKnown) {
			LOG.debug("Removing terminal {}.", next.getIFDName());
			IFDStatusType newStatus = new IFDStatusType();
			newStatus.setIFDName(next.getIFDName());
			newStatus.setConnected(Boolean.FALSE);
			result.add(newStatus);
		    }
		}
		return result;
	    } else {
		throw ex;
	    }
	}
    }

    /**
     * Resets a card given as connection handle.
     *
     * @param cHandleRm {@link ConnectionHandleType} object representing a card which shall be removed.
     * @param cHandleIn {@link ConnectionHandleType} object representing a card which shall be inserted.
     * @param ifaceProtocol Interface protocol of the connected card.
     */
    public void resetCard(ConnectionHandleType cHandleRm, ConnectionHandleType cHandleIn, String ifaceProtocol) {
	env.getEventDispatcher().notify(EventType.CARD_REMOVED, new IfdEventObject(cHandleRm, null));

	// determine if the reader has a protected auth path
	IFDCapabilitiesType slotCapabilities = getCapabilities(cHandleRm.getContextHandle(), cHandleRm.getIFDName());
	boolean protectedAuthPath = slotCapabilities != null ? ! slotCapabilities.getKeyPadCapability().isEmpty() : false;

	HandlerBuilder chBuilder = HandlerBuilder.create();
	ConnectionHandleType cInNew = chBuilder.setSessionId(sessionId)
		.setCardType(cHandleIn.getRecognitionInfo())
		.setCardIdentifier(cHandleIn.getRecognitionInfo())
		.setContextHandle(cHandleIn.getContextHandle())
		.setIfdName(cHandleIn.getIFDName())
		.setSlotIdx(BigInteger.ZERO)
		.setSlotHandle(cHandleIn.getSlotHandle())
		.setProtectedAuthPath(protectedAuthPath)
		.buildConnectionHandle();
	env.getEventDispatcher().notify(EventType.CARD_INSERTED, new IfdEventObject(cInNew));

	if (isRecognize()) {
	    Recognizer rec = new Recognizer(env, cInNew, ifaceProtocol);
	    Thread recThread = new Thread(rec, "Recoginiton-" + THREAD_NUM.getAndIncrement());
	    recThread.start();
	}
    }

    boolean isRecognize() {
	return env.getRecognition() != null;
    }

    @Nullable
    private IFDCapabilitiesType getCapabilities(byte[] ctxHandle, String ifdName) {
	GetIFDCapabilities req = new GetIFDCapabilities();
	req.setContextHandle(ctxHandle);
	req.setIFDName(ifdName);
	GetIFDCapabilitiesResponse res = (GetIFDCapabilitiesResponse) env.getDispatcher().safeDeliver(req);
	return res.getIFDCapabilities();
    }

}

File: ifd/ifd-core/src/jvmMain/kotlin/org/openecard/ifd/event/IfdEventManager.kt
/****************************************************************************
 * Copyright (C) 2012-2019 ecsec GmbH.
 * All rights reserved.
 * Contact: ecsec GmbH (info@ecsec.de)
 *
 * This file is part of the Open eCard App.
 *
 * GNU General Public License Usage
 * This file may be used under the terms of the GNU General Public
 * License version 3.0 as published by the Free Software Foundation
 * and appearing in the file LICENSE.GPL included in the packaging of
 * this file. Please review the following information to ensure the
 * GNU General Public License version 3.0 requirements will be met:
 * http://www.gnu.org/copyleft/gpl.html.
 *
 * Other Usage
 * Alternatively, this file may be used in accordance with the terms
 * and conditions contained in a signed written agreement between
 * you and ecsec GmbH.
 *
 ***************************************************************************/
package org.openecard.ifd.event

import io.github.oshai.kotlinlogging.KotlinLogging
import iso.std.iso_iec._24727.tech.schema.*
import org.openecard.common.ECardConstants
import org.openecard.common.WSHelper
import org.openecard.common.WSHelper.checkResult
import org.openecard.common.event.EventType
import org.openecard.common.event.IfdEventObject
import org.openecard.common.interfaces.Environment
import org.openecard.common.util.HandlerBuilder
import org.openecard.common.util.ValueGenerators.genBase64Session
import java.lang.Boolean
import java.math.BigInteger
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.Future
import java.util.concurrent.ThreadFactory
import java.util.concurrent.atomic.AtomicInteger
import kotlin.ByteArray
import kotlin.RuntimeException
import kotlin.String


private val LOG = KotlinLogging.logger {  }

/**
 * Main class of the event system.
 * Use this to create and operate an event manager.
 *
 * @author Tobias Wich
 * @author René Lottes
 */
class IfdEventManager(
	protected val env: Environment,
	protected val ctx: ByteArray
) {

    protected val sessionId: String = genBase64Session()
	private val builder: HandlerBuilder = HandlerBuilder.create()
.setContextHandle(ctx)
.setSessionId(sessionId)

	protected var threadPool: ExecutorService? = null

    private var eventRunner: IfdEventRunner? = null
    private var watcher: Future<*>? = null

	@Synchronized
    fun initialize() {
        threadPool = Executors.newCachedThreadPool(object : ThreadFactory {
            private val num = AtomicInteger(0)
            private val group = ThreadGroup("IFD Event Manager")
            override fun newThread(r: Runnable): Thread {
                val name = String.format("IFD Watcher %d", num.getAndIncrement())
                val t = Thread(group, r, name)
                t.setDaemon(false)
                return t
            }
        })
        // start watcher thread
        try {
            eventRunner = IfdEventRunner(env, this, builder, ctx)
            watcher = threadPool!!.submit(eventRunner!!)
        } catch (ex: WSHelper.WSException) {
            throw RuntimeException("Failed to request initial status from IFD.")
        }
    }

    @Synchronized
    fun terminate() {
        eventRunner!!.setStoppedFlag()
        watcher!!.cancel(true)
        threadPool!!.shutdownNow()
    }

    @Throws(WSHelper.WSException::class)
    fun wait(lastKnown: List<IFDStatusType>): List<IFDStatusType> {
        val wait = Wait()
        wait.setContextHandle(ctx)
        wait.getIFDStatus().addAll(lastKnown)
        val resp = env.ifd!!.wait(wait)

        try {
            checkResult(resp)
            val result = resp.getIFDEvent()
            return result
        } catch (ex: WSHelper.WSException) {
            if (ECardConstants.Minor.IFD.INVALID_SLOT_HANDLE == ex.resultMinor) {
                // this can only happen when the PCSC stack is reloaded, notify all cards have disappeared
                val result = mutableListOf<IFDStatusType>()
                if (!lastKnown.isEmpty()) {
					LOG.info { "PCSC stack seemed to disappear. Signalling that no cards are available anymore." }
                    for (next in lastKnown) {
						LOG.debug { "Removing terminal ${next.getIFDName()}." }
                        val newStatus = IFDStatusType()
                        newStatus.setIFDName(next.getIFDName())
						newStatus.isConnected = Boolean.FALSE
                        result.add(newStatus)
                    }
                }
                return result
            } else {
                throw ex
            }
        }
    }

    /**
     * Resets a card given as connection handle.
     *
     * @param cHandleRm [ConnectionHandleType] object representing a card which shall be removed.
     * @param cHandleIn [ConnectionHandleType] object representing a card which shall be inserted.
     * @param ifaceProtocol Interface protocol of the connected card.
     */
    fun emitResetCardEvent(cHandleRm: ConnectionHandleType, cHandleIn: ConnectionHandleType, ifaceProtocol: String?) {
        // determine if the reader has a protected auth path
        val slotCapabilities = getCapabilities(cHandleRm.getContextHandle(), cHandleRm.getIFDName())
        val protectedAuthPath =
            if (slotCapabilities != null) !slotCapabilities.getKeyPadCapability().isEmpty() else false

        val chBuilder = HandlerBuilder.create()
        val cInNew = chBuilder.setSessionId(sessionId)
            .setCardType(cHandleIn.getRecognitionInfo())
            .setCardIdentifier(cHandleIn.getRecognitionInfo())
            .setContextHandle(cHandleIn.getContextHandle())
            .setIfdName(cHandleIn.getIFDName())
            .setSlotIdx(BigInteger.ZERO)
            .setSlotHandle(cHandleIn.getSlotHandle())
            .setProtectedAuthPath(protectedAuthPath)
            .buildConnectionHandle()
        env.eventDispatcher!!.notify(EventType.CARD_RESET, IfdEventObject(cInNew, ifaceProtocol, true))
    }

    private fun getCapabilities(ctxHandle: ByteArray?, ifdName: String?): IFDCapabilitiesType? {
        val req = GetIFDCapabilities()
        req.setContextHandle(ctxHandle)
        req.setIFDName(ifdName)
        val res = env.dispatcher!!.safeDeliver(req) as GetIFDCapabilitiesResponse
        return res.getIFDCapabilities()
    }

}

private val THREAD_NUM = AtomicInteger(1)
